articles/external_qos_file.md

```diff
@@ -0,0 +1,402 @@
+---
+layout: default
+title: Configuring QoS from an external file
+permalink: articles/qos_file.html
+abstract:
+  This article describes how to configure QoS from an external file.
+author: '[Ivan Santiago Paunovic](https://github.com/ivanpauno)'
+published: true
+---
+
+- This will become a table of contents (this text will be scraped).
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Overview
+
+`ROS 2` added the [Quality of Service (QoS)](qos.md) to publishers, subscriptions, clients and services.
+Up `ROS 2 Foxy`, the `QoS` can only be specified in code.
+To avoid recompiling the source code with patched `QoS`, node implementers have used different mechanisms to configure `QoS` profiles:
+- Command line arguments
+- ROS parameters
+- Combining system default QoS with vendor specific `QoS` profiles.
+
+Instead of having different ways of externally specifying `QoS` for the different endpoints, it would be good to have a standardized way to specific `QoS` profiles.
+
+## QoS file format
+
+DDS uses the concept of QoS profiles ([rti documentation](https://community.rti.com/examples/using-qos-profiles)), in which each profile has a name. When creating an entity (e.g. a `DataWriter`), the passed profile name will be looked up in the qos profiles file being used, and the QoS can be loaded in that way.
+
+ROS 2 can use a simpler approach, leveraging node name uniqueness <sup id="back_node_name_uniqueness">[1](#to_node_name_uniqueness)</sup>.
+
+<b id="to_node_name_uniqueness">1</b> Node name uniqueness is not being enforced up to Foxy, though it is supposed. [↩](#back_node_name_uniqueness)
+
+### Nodes
+
+In an acceptable `QoS` file format, it should be to distinguish `QoS` settings targeted to different nodes.
+Thus, a top level tag to identify the node should be available.
+Example alternatives:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        ...
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        ...
+```
+
+### Endpoint type
+
+It should be possible to distinguish between the different endpoint types in a qos profile file.
+For example:
+
+```xml
+<qos_profiles>
+    <node name="my_node" namespace="/my_ns/nested_ns">
+        <publisher topic_name="asd">  <!--Relative name-->
+            ...
+        </publisher>
+        <subscription topic_name="/another_ns/asd">  <!--Absolute name-->
+            ...
+        </subscription>
+    </node>
+<qos_profiles>
+```
+
+```yaml
+/my_ns/nested_ns/my_node:
+    ros__qos_profiles:
+        publisher:
+            topic_name: asd
+            qos:
+                ...
+        subscription:
+            topic_name: /another_ns/asd
+            qos:
+                ...
+```
+
+Besides `publisher` and `subscription` tags, `client` and `service` should be allowed.
```

Comment:
Has there been any progress on figuring out QoS for actions?

Reply:
AFAIK, users are able to configure the QoS of the separate topics and services that make up an action, but I don't think there's been any discussion about applying a QoS profile to an action as a whole.

So, in theory, we should be able to specify QoS overrides for the topics and services for an action but I'm not sure what it would look like (or how meaningful it would be).

Reply:
Yeah, currently action QoS can be configured for each topic and service that make up the action (see [here](https://github.com/ros2/rcl/blob/af438bcca1ffb69714c0b225ba0f80aceab04906/rcl_action/include/rcl_action/action_client.h#L44-L52)).
I haven't thought if that was a good idea or not, maybe we should have only one action QoS, instead of five.

I think the proposed file format should include actions, so I will think on how they can be included.

---

Unrelated note: We currently have only on `rmw_qos_profile_t` type, which I don't think is a good idea.
For example, the `deadline` QoS is ignored in subscriptions.

Maybe we should clean that up before implementing this proposal, though it doesn't matter much now.

Reply:
> Maybe we should clean that up before implementing this proposal, though it doesn't matter much now.

Sounds like scope creep; opening an issue for that clean up would be alright, but I guess it can happen independent of this feature.

Reply:
> Sounds like scope creep

I have an unfortunate tendency to creep the scope of what I'm doing a lot :joy:.

Reply:
I have added a note about actions in https://github.com/ros2/design/pull/296/commits/274584b12c04eaaab72c2377fa4a1baf77b47338.